lib/spack/spack/version/git_ref_lookup.py

```diff
@@ -93,14 +94,16 @@ def fetcher(self):
     @property
     def repository_uri(self):
         """Identifier for git repos used within the repo and metadata caches."""
-        try:
-            components = [
-                str(c).lstrip("/") for c in spack.util.url.parse_git_url(self.pkg.git) if c
-            ]
-            return os.path.join(*components)
-        except ValueError:
-            # If it's not a git url, it's a local path
-            return os.path.abspath(self.pkg.git)
+        (scheme, netloc, path, params, query, fragment) = urllib.parse.urlparse(self.pkg.git)
+        path = path.lstrip("/")
+        if scheme == "file":
+            unique_dir = Path(path)
+            if sys.platform == "win32":
+                unique_dir = spack.util.hash.b32_hash(Path(path).as_posix())[-7:]
+            return Path("file") / unique_dir
+
+        (hostname, port) = netloc.split(":")
```

Comment:
If netloc does not contain a port, this will fail.

Also, 

```
>>> parse.urlparse("git@github.com:spack/spack.git")
ParseResult(scheme='', netloc='', path='git@github.com:spack/spack.git', params='', query='', fragment='')
```

the original implementation did some extra work to remove `@` and `:` from this (overall, urlparse doesn't appear to break up this git-based URL into components).

Also, long URLs could trigger the same path-length issues we'd worry about with files. So we'd probably want to hash everywhere.

It occurs to me we could avoid all this parsing confusion by hashing everywhere (vs. we discussed only doing so on Windows). So like:

```
spack.util.hash.b32_hash(self.pkg.git)
```

And another option is to keep the logic (i.e. keep `url.parse_git_url`) and then hash on Windows like:

```
       try:
            components = [
                str(c).lstrip("/") for c in spack.util.url.parse_git_url(self.pkg.git) if c
            ]
            path = os.path.join(*components)
        except ValueError:
            # If it's not a git url, it's a local path
            path = os.path.abspath(self.pkg.git)
    if win32:
        return b32_hash(path)
    else:
        return path
```